README
Project Clock
=============
$Id: README,v 1.7 2004/03/17 22:11:36 david Exp $

Changes:
	$Log: README,v $
	Revision 1.7  2004/03/17 22:11:36  david
	changed for 0.7

	Revision 1.6  2001/01/24 03:21:51  david
	added features for v0.6

	Revision 1.5  2000/06/07 02:05:05  david
	added summary usage

	Revision 1.4  2000/02/28 05:26:55  david
	version 0.4

	Revision 1.3  2000/02/23 22:24:07  david
	v0.3 updates

	Revision 1.2  2000/02/23 22:17:39  david
	v0.2 changes

	Revision 1.1  2000/02/01 23:14:28  david
	Initial revision


Version 0.7

David Keeffe; david@systemsolve.net
---------------------------------------

I often work on several different projects at once, and really need
to keep track of time on each so I can bill clients correctly and fairly.

Noting the time in a diary was a start, but if your desk is anything
like mine, finding the diary *and* a writing implement is a stressful
activity. So the time gets recorded occasionally and inaccurately.

The only constant is my Linux desktop, so I looked for a simple program
which would let me run timers for multiple projects, where a click of
the mouse would stop one timer and start another.

I wrote this in Tcl/Tk. It just about does enough for me;
it's a bit short on reporting tools (but see "summary" below).

The version as reported by "Help->About" is 1.x - this is because RCS starts
at 1.1!

Read the CHANGES file!

Copyright
---------

This program is COPYRIGHT 2000-2001 David Keeffe. You may use and distribute
it under the terms of the GNU Public License.

Installing Project Clock
------------------------

Read the INSTALL file in this directory.

Using Project Clock
-------------------

Start Project Clock by issuing the command "pck-timer".

Pclock will create and maintain a project record automatically.

When you fire up pck-timer, there will be no projects. Use the menu item
'Edit->Add Project' to add an item, and 'Edit->Delete Project' to delete
an item.

A project timer will start when you click on the project button with the
left mouse button.

You can update times manually by RIGHT-clicking on the project button.

Every two minutes, the project state will be written out to a file
'$HOME/.pclock', with a backup to '$HOME/.pclock.bak'.

The data is saved when you exit and when you send the program SIGINT.

The data is formatted as a couple of Tcl associative arrays.

(new in v0.7)

You can keep a simple diary of work for a project. If you click the
"quill" icon on the toolbar, a window will pop up with diary entries
for the currenly selected project, and space to write a new one. If no
project is current, a list will be popped up first.

Diary entries normally go in $HOME/pclock.d/diary.<project-name>. It's
possible to implement a shared diary by setting the diary home to be a
world- (or group-) writable directory, but note that there is no locking
presently implemented.

(new in v0.6)

You can display a number of reports from pck-timer.
The reports are generated using pck-summary (see below). Currently pck-timer
calls pck-summary using "exec". This won't work on MS Windows. There is a
function "runprog" in the utilproc.tcl library, but I haven't tested it
with Project Clock yet.

Various properties are saved in a file $HOME/.pclockrc and can be set
using the built-in properties editor. You can also edit the rc file by
hand.

Report generator "summary"
--------------------------

You can generate a summary for a given range of dates, using the "pck-summary"
program. You can also specify one or more projects to restrict output.
Also, because of the way Tcl works, you can specify a project name using
a "glob" style pattern. This feature is a happy accident!

Usage: pck-summary [-hxlrs][-f day][-t day][-d range][project-name ...]
Where -
	-h            : output this message
	-x            : increment debug level
	-l            : list projects
	-q            : don't print headings
	-r            : just rawtimes times, no factoring
	-s            : omit summaries
	-v            : print daily times
	-f            : start day (default 1)
	-t            : end day (default last of month)
	-d period     : process times for 'period' (single (m/y) or range (m/y-m/y)) 
	-i file       : read data from 'file'



CHANGES
Wed Dec 25 17:39:13 CET 2019

  - now converted to a starkit / starpack for single executable deployment
  - one can provide a working directory on the command line, so data storage location is settable
  - one can now start severeral instances at the same time, if they work on different working directories
    (else data loss may occur)
  - default project for startup is selectable
  - some more reports from menu available
  - data update to disk is performed on stopping all clocks

Tue Sep  6 08:52:43 EST 2005

projclock.tcl
	- changed name from Project Clock to WorkerTimer due to trademark issue

Wed Mar 17 08:40:14 EST 2004

projclock.tcl
    - some tidying up of window behaviour
	- fixed problem with custom summary when pclock dir has spaces in name

Tue Mar  6 09:42:05 EST 2001

my birthday!

projclock.tcl
	- added history list and state file; scope for other state changes
		state file is separate from rc file
	- considering next relase

Fri Mar  2 14:50:37 EST 2001

projclock.tcl
	- added method for having different time database files

summary.tcl
	- simplified main loop
	- allowed days in date range option
	- allow reporting in fractional hours

Mon Feb  5 11:04:15 EST 2001

projclock.tcl
	- added diary system
	- changed GUI summary system not to use separate process

Wed Jan 24 13:57:33 EST 2001

projclock.tcl
	- added View menus to view summaries created by pck-summary
	- added summary viewer with save option
	- added properties editor

summary.tcl
	- added date range capability

Wed Jun  7 11:52:22 EST 2000

Changed wrappers to allow call of devel copy.

projclock.tcl 
	- Changed save mode to drop entries with 0 time.

Fri Mar 17 11:53:52 EST 2000

RPM now available as well as tgz.

projclock.tcl
	- the end-of-day code was broken; it's fixed now


Fri Feb 25 14:51:32 EST 2000

IMPORTANT!!!!
	The startup wrapper has been renamed from pclock to pck-timer!
	The summariser is now started as pck-summary!
IMPORTANT!!!!

projclock.tcl
	- having left a timer running overnight (again) I implemented an
	  "end-of-day" watchdog, which gives you 5 minutes to continue or
	  have the current timer stopped automatically.
	- added hooks for a ~/.pclockrc file (config tools coming later)

wrappers
	- created one for summary, to avoid the missing included files
	- changed the name from pclock to pck-timer, summary to pck-summary
	- make install now installs these; make "oinstall" installs using the
	  old names

Thu Feb 24 09:12:58 EST 2000

projclock.tcl
	- changed timer cycle to use one "after", with a check for save cycles
	- changed redraw strategy to try to eliminate crashes; it seems that
	  hundreds of 'grid forget' calls causes an internal overflow

summary.tcl
	- made useful: run summary -h to see what you can do

Wed Feb  2 10:05:49 EST 2000

utilproc.tcl
	- added numeric filter version of EntryBox
	- added RCS log info
	- PopupWindow convenience fcn.
	- Info convenience fcn.

projclock.tcl
	- added RCS Log info
	- tooltips
	- hooks for property sheet
	- hooks for configurable cycle timers
	- timer increment now uses wallclock time and not a simple
	  counter
	- added (very short) about and help
	- cumulative time now based on 8-hour (cycles(workday)) day
	- small cleanup on how timers are started

Makefile 
	created!


projclock.vfs/lib/app-projclock/summary.tcl
#!/usr/bin/tcl
#
# Summarising module for Project Clock
#
# David Keeffe
# Raster Solutions P/L
#
# $Id: summary.tcl,v 1.4 2004/03/17 22:14:47 david Exp $

# History:
#	$Log: summary.tcl,v $
#	Revision 1.4  2004/03/17 22:14:47  david
#	changed date format for -d
#	added -T option for hours only
#
#	Revision 1.3  2001/01/24 03:20:59  david
#	added summary date ranges
#	added multiple projects
#
#	Revision 1.2  2000/06/07 02:01:10  david
#	lots of options
#
#	Revision 1.1  2000/02/21 23:07:03  david
#	Initial revision
#
#

set progname [file tail $argv0]
set progdir [file dirname $argv0]
set env(LANG) "C"
set env(LC_ALL) "C"

source "$progdir/getopt.tcl"

if {[array names env PCLOCK_DATADIR] != ""} then {
  set datadir $env(PCLOCK_DATADIR)
} else {
  set datadir $env(HOME)
}

set dbfile "$datadir/.pclock"
set pconfig "$datadir/.pclockrc"
set projects *
set listonly false
set summarise 1
set verbose 0
set quiet 0
set day 1
set eday end
set year {}
set month {}
set eyear {}
set emonth {}
set debug 0
set workday 8
set rawtimes 0
set showhours 0

set usagelist {
	{h {} {output this message}}
	{T {} {show times as hours}}
	{x {} {increment debug level}}
	{l {} {list projects}}
	{q {} {don't print headings}}
	{r {} {just rawtimes times, no factoring}}
	{s {} {omit summaries}}
	{v {} {print daily times}}
	{d {period} {process times for 'period' (single ([d/]m/y) or range ([d/]m/y-[d/]m/y)) }}
	{i {file} {read data from 'file'}}
}


set ufmt {Usage: %s [-hTxlrs][-d range][project-name]}
	

proc usage { { fd stderr }} {
	global argv0 usagelist ufmt
	puts $fd [format $ufmt $argv0]
	puts $fd "Where -"
	foreach l $usagelist {
		puts $fd [format {	-%s %-10s : %s} [lindex $l 0] [lindex $l 1] [lindex $l 2]]
	}
}

proc getdays { date plist } {
	global days

	set result {}
	# puts stderr "getdays: $date -> $plist"

	foreach p $plist {
		foreach n [array names days "$date,$p"] {
			if {[lsearch -exact $result $n] == -1} {
				lappend result $n
			}
		}
	}
	return $result
}


proc timeof { value showdays } {
	global workday showhours
	if { $showhours } {
		if { $value > 0 } {
			set hours [expr floor(10 * (($value+300.0) / ( 60.0 * 60.0 )))/10.0]
		} else {
			set hours 0.0
		}
		return "    [format %#5.1f $hours]"
	} elseif { $showdays } {
		set days [expr $value / ( $workday * 60 * 60 )]
		set hours [expr $value % ( $workday * 60 * 60 )]
		return "[format %2dd $days] [clock format $hours -format {%H:%M:%S} -gmt 1]"
	} else {
		return "    [clock format $value -format {%H:%M:%S} -gmt 1]"
	}
}


while { [ set err [getopt $argv "THhdrxqlsvf:t:i:d:" opt arg] ] } {
	if { $err <  0 } {
		puts stderr "$argv0: $arg"
	} else {
		switch -exact $opt {
		T {
			set showhours 1
		}

		H {
			usage stdout
			exit 0
		}

		h {
			usage
			exit 0
		}

		d {
			set datexx [split $arg -]
			
			if {[llength $datexx] == 1} {
				set dstr [lindex $datexx 0]
				set date [split $dstr /]
				set d0 [lindex $date 0]
				set d1 [lindex $date 1]
				set d2 [lindex $date 2]
				if {[string match -nocase {[a-z]*} $d0]} {
					set day 1
					set month $d0
					set year $d1
				} else {
					set day $d0
					set month $d1
					set year $d2
				}
				set emonth {}
				set eyear {}
			} else {
				set day {1}
				set month {epoch}
				set year {epoch}
				set eday {end}
				set emonth {eot}
				set eyear {eot}
				if {[lindex $datexx 0] != {} } {
					set dstr [lindex $datexx 0]
					set date [split $dstr /]
					set d0 [lindex $date 0]
					set d1 [lindex $date 1]
					set d2 [lindex $date 2]
					if {[string match -nocase {[a-z]*} $d0]} {
						set day 1
						set month $d0
						set year $d1
					} else {
						set day $d0
						set month $d1
						set year $d2
					}
				}
				if {[lindex $datexx 1] != {} } {
					set dstr [lindex $datexx 1]
					set date [split $dstr /]
					set d0 [lindex $date 0]
					set d1 [lindex $date 1]
					set d2 [lindex $date 2]
					if {[string match -nocase {[a-z]*} $d0]} {
						set eday end
						set emonth $d0
						set eyear $d1
					} else {
						set eday $d0
						set emonth $d1
						set eyear $d2
					}
				}
			}

		}

		r {
			set rawtimes 1
		}

		i {
			set dbfile $arg
		}

		x {
			incr debug
		}

		l {
			set listonly true
		}

		q {
			set quiet 1
		}

		v {
			incr verbose
		}

		s {
			set summarise 0
		}

		}
	}
}

if { $debug > 1 } {
	cmdtrace on stderr
}

set argv [lrange $argv $optind end] 

set projects $argv

if { $projects == {} } {
	set projects *
}

if {[file readable $pconfig]} {
	source $pconfig
}


set now [clock seconds]

if { $month == {epoch} } {
	set month Jan
}

if { $year == {epoch} } {
	set year [clock format $now -format "%Y" -gmt 1]
}

if { $month == {} } {
	set month [clock format $now -format "%b" -gmt 1]
}

if { $year == {} } {
	set year [clock format $now -format "%Y" -gmt 1]
}

if { $emonth == {eot} } {
	set emonth Dec
}

if { $eyear == {eot} } {
	set eyear [clock format $now -format "%Y" -gmt 1]
}

if {[catch {source $dbfile} res]} {
	puts stderr "$res"
	exit 1
}

if { $listonly } {
	foreach n [array names days] {
		regsub {[-a-zA-Z0-9]*,(.*)} $n {\1} p
		set plist($p) 1
	}
	puts [join [array names plist] "\n"]
	exit 0
}

array set months_n {
Jan 1
Feb 2
Mar 3
Apr 4
May 5
Jun 6
Jul 7
Aug 8
Sep 9
Oct 10
Nov 11
Dec 12
}

foreach m [array names months_n] {
	set n_months($months_n($m)) $m
}

array set alldays {}
set total 0

if [scan $day "%d" t] {
	set day $t
	unset t
}

if { $emonth != {} || $eyear != {} } {
	if { $eday == "end" } { 
		set tbanner [format "Times from %04d-%02d-%02d to %s" $year $months_n($month) $day $eday]
	} elseif [scan $eday "%d" t] {
		set eday $t
		unset t
		set tbanner [format "Times from %04d-%02d-%02d to %04d-%02d-%02d" $year $months_n($month) $day $eyear $months_n($emonth) $eday]
	} else {
		set tbanner [format "Times from %04d-%02d-%02d to %4.4s-%2.2s-%2.2s" $year $months_n($month) $day $eyear $months_n($emonth) $eday]
	}
} else {
	set tbanner [format "Times for %04d-%02d" $year $months_n($month)]
}
if { $verbose && !$quiet } {
	puts $tbanner
	puts [format {%-15s %-20s %8s} Date Project Time]
	puts [format {%-15s-%-20s-%s} --------------- --------------------- -----------]
}

if { $emonth == {} } {
	set emonth $month
}

if { $eyear == {} } {
	set eyear $year
}


# puts stderr "FROM $day $month $year TO $eday $emonth $eyear"
set dstart [clock scan "$day $month $year" -gmt 1]
set daysec [expr 24*60*60]
if { $eday == "end" } {
	set dend [expr [clock scan "1 $emonth $eyear + 1 month" -gmt 1] - $daysec]
} else {
	set dend [clock scan "$eday $emonth $eyear" -gmt 1]
}

if { $dstart >= $dend } {
	puts stderr "$progname: start is before end!"
	exit 0
}


# puts stderr "START \[2\]: [clock format $dstart -gmt 1] - END: [clock format $dend -gmt 1]"

for { set actual $dstart} { $actual <= $dend } { incr actual $daysec } {

	# puts stderr "ACTUAL: [clock format $actual -gmt 1]"

	set date [clock format $actual -format "%d-%b-%Y" -gmt 1]
	set fdate [clock format $actual -format "%a %Y-%m-%d" -gmt 1]
	set matches [lsort [getdays $date $projects]] 
	if { ($verbose > 1) && $matches == {} } {
		puts [format {%s %-20s %s} $fdate "--EMPTY--" 0]
	}
	foreach n $matches {
		regsub "$date," $n {} proj
		if { $rawtimes == 0 && [info exists factor($proj)] } {
			set dd [expr int($factor($proj) * $days($n))]
		} else {
			set dd $days($n)
		}
		if {[info exists alldays($proj)]} {
			incr alldays($proj) $dd
		} else {
			set alldays($proj) $dd
		}
		incr total $dd
		if { $verbose > 1 || ( $verbose && $dd > 0 ) } {
			puts [format {%s %-20s %s} $fdate $proj [timeof $dd 0]]
		}
	}
}

if { $summarise } {
	if { !$quiet } {
		if { !$verbose } {
			puts $tbanner
			puts [format {%-15s %-20s %8s} {} Project Time]
		}
		puts [format {%-15s-%-20s-%s} --------------- --------------------- -----------]
	}
	foreach n [lsort [array names alldays]] {
		puts [format {%-15s %-20s %s} {} $n [timeof $alldays($n) 1]]	
	}
	if { !$quiet } {
		puts [format {%-15s-%-20s-%s} --------------- --------------------- -----------]
		puts [format {%-15s %-20s %s} {} TOTAL [timeof $total 1]]
	}
}
